### PR TITLE
Finalize context if and only if it's shutdown

### PR DESCRIPTION
**Pull request by hidmic**
This pull request, albeit controversial, ensures `rmw_context_fini()` fails on a not shutdown context, as its API contract establishes (see https://github.com/ros2/rmw/pull/242).

### CODE REVIEW

File: rmw_connext_cpp/src/rmw_init.cpp [hunk 1]
```diff
@@ -105,12 +110,27 @@ rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
...
+  context->impl->is_shutdown = false;
   rmw_ret_t ret = rmw_init_options_copy(options, &context->options);
   if (RMW_RET_OK != ret) {
+    delete context->impl;
     return ret;
   }
-  return init();
+  ret = init();
+  if (RMW_RET_OK != ret) {
+    if (RMW_RET_OK != rmw_init_options_fini(&context->options)) {
```
Comment (ivanpauno) [inline rmw_connext_cpp/src/rmw_init.cpp:126]: this seem unrelated to the PR.
IMO, it should be done in another PR, or in a separate commit.

Comment (hidmic) [reply rmw_connext_cpp/src/rmw_init.cpp:126]: That is fair, yeah. I came across leaks and I couldn't help it.

**Comment by hidmic** [comment]
CI up to `test_rmw_implementation` against all RMW implementations:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11240/) (canary build)

**Comment by ivanpauno** [comment]
> CI up to test_rmw_implementation against all RMW implementations:

I would run test including up to an upper layer, like test_rclcp, just in case.
At least in one platform.
